### PR TITLE
Externalize samples testing infrastructure

### PR DESCRIPTION
**Pull request by eriwen**
This change replaces UserGuideSamplesRunner with an external library
called sample-check. Sample discovery, execution, and normalization
is delegated to this library, with some custom normalizers remaining
in the Gradle codebase.

Samples output has become co-located with the sample code. ~~Samples tests
must be run with Java 8+.~~ Samples tests can now be run with Java 7 due to recent changes in exemplar.

Summary of changes:
 - Moved samples output alongside sample projects
 - Removed UserGuideSamplesRunner
 - Creating new OutputNormalizers
 - Use GradleSamplesRunner and new normalizers for UserGuideSamplesIntegrationTest

NOTE: Samples for the Gradle wrapper have been disabled since #3569, and so this is more formalized by omitting sample testing config for them. 

You will also be interested in https://github.com/gradle/exemplar/, which is to be shared publicly after some polishing.

### Context
Once this is complete, we can convert docbook passthrough `<sample>`s to asciidoc syntax, which will allow the user manual to support switchable Groovy/Kotlin snippets, as well as be much faster and eliminate a bunch of custom stuff we're doing for docs processing/testing.

### Gradle Core Team Checklist
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation
- [ ] Recognize contributor in release notes

### CODE REVIEW

File: subprojects/docs/src/samples/native-binaries/cunit/completeCUnitExample.out [hunk 1]
```diff
@@ -0,0 +1,5 @@
+
+There were test failures:
+  1. /home/user/gradle/samples/src/operatorsTest/c/test_plus.c:6  - plus(0, -2) == -2
+  2. /home/user/gradle/samples/src/operatorsTest/c/test_plus.c:7  - plus(2, 2) == 4
+
```
Comment (adammurdoch) [inline subprojects/docs/src/samples/native-binaries/cunit/completeCUnitExample.out]: did you pick up the changes that we made to the samples test runner in `gradle/gradle` that merges stdout and stderr, so the that failure details and so on are also included in the output (and can be verified)? It looks like this is missing given this output file.

File: subprojects/docs/src/samples/userguide/buildlifecycle/buildProjectEvaluateEvents/settings.gradle [hunk 2]
```diff
@@ -1,3 +1,5 @@
+rootProject.name = 'buildProjectEvaluateEvents'
```
Comment (adammurdoch) [inline subprojects/docs/src/samples/userguide/buildlifecycle/buildProjectEvaluateEvents/settings.gradle]: why is this needed?

File: subprojects/docs/src/samples/userguide/tasks/addToTaskContainer/addToTaskContainer.sample.conf [hunk 3]
```diff
@@ -0,0 +1,3 @@
+executable: gradle
+args: tasks
+expect-failure: false
```
Comment (adammurdoch) [inline subprojects/docs/src/samples/userguide/tasks/addToTaskContainer/addToTaskContainer.sample.conf]: seems like there's a lot of duplication in these .conf files.

**Comment by adammurdoch** [comment]
This is great, looking forward to using this for the native samples.

File: subprojects/docs/src/samples/userguide/tutorial/configureTaskUsingProjectProperty/configureTaskUsingProjectProperty.sample.conf [hunk 4]
```diff
@@ -0,0 +1,10 @@
+# tag::cli[]
+# gradle performRelease -PisCI=true --quiet
```
Comment (acourtneybrown) [inline subprojects/docs/src/samples/userguide/tutorial/configureTaskUsingProjectProperty/configureTaskUsingProjectProperty.sample.conf]: move `args` to end?

File: subprojects/docs/src/samples/userguide/tutorial/excludeTasks/excludeTask.sample.conf [hunk 5]
```diff
@@ -0,0 +1,10 @@
+# tag::cli[]
+# gradle dist --exclude-task=test
```
Comment (acourtneybrown) [inline subprojects/docs/src/samples/userguide/tutorial/excludeTasks/excludeTask.sample.conf]: move `dist` to end?

File: subprojects/internal-integ-testing/src/main/groovy/org/gradle/integtests/fixtures/executer/SequentialOutputMatcher.groovy [hunk 6]
```diff
@@ -18,19 +18,18 @@
...
-
 /**
  * Check that the actual output lines match the expected output lines in content and order.
  */
 class SequentialOutputMatcher {
     private static final String NL = SystemProperties.instance.lineSeparator
 
     public void assertOutputMatches(String expected, String actual, boolean ignoreExtraLines) {
-        List actualLines = normaliseOutput(actual.readLines()).findAll { !it.isEmpty() }
+        List actualLines = new ArtifactResolutionOmittingOutputNormalizer().normalize(actual).readLines().findAll { !it.isEmpty() }
```
Comment (acourtneybrown) [inline subprojects/internal-integ-testing/src/main/groovy/org/gradle/integtests/fixtures/executer/SequentialOutputMatcher.groovy]: nit: make a singleton for `ArtifactResolutionOmittingOutputNormalizer`?

File: subprojects/docs/src/samples/java/multiproject/settings.gradle [hunk 7]
```diff
@@ -1,3 +1,4 @@
+rootProject.name = 'multiproject'
```
[the remark this answers is not in the document]
Comment (eriwen) [reply subprojects/docs/src/samples/java/multiproject/settings.gradle]: This may be due to the way sample-check copies sample projects to a temporary location. If I omit this, then the `listProjects` sample fails with:

> Expected: Root project 'multiproject'
Actual: Root project 'java_multiproject_listProjects.sample'

File: subprojects/docs/src/samples/userguide/buildlifecycle/buildProjectEvaluateEvents/settings.gradle (hunk 2 above)
Comment (eriwen) [reply subprojects/docs/src/samples/userguide/buildlifecycle/buildProjectEvaluateEvents/settings.gradle]: This is similar to above, to allow output verification to match the project name.

File: subprojects/docs/src/samples/userguide/tasks/addToTaskContainer/addToTaskContainer.sample.conf (hunk 3 above)
Comment (eriwen) [reply subprojects/docs/src/samples/userguide/tasks/addToTaskContainer/addToTaskContainer.sample.conf]: I have removed all redundant lines.

File: subprojects/docs/src/samples/native-binaries/cunit/completeCUnitExample.out (hunk 1 above)
Comment (eriwen) [reply subprojects/docs/src/samples/native-binaries/cunit/completeCUnitExample.out]: Bummer, I did not. I'll fix this up, and hope I can merge quickly so we don't lose other changes like this.

File: subprojects/docs/src/samples/java/multiproject/settings.gradle (hunk 7 above)
[the remark this answers is not in the document]
Comment (adammurdoch) [reply subprojects/docs/src/samples/java/multiproject/settings.gradle]: this is a problem in exemplar, then. It should preserve the directory name when creating the copy.

File: subprojects/docs/src/samples/ivy-publish/java-multi-project/publishing_ivy:publish-custom-artifact-snippet.sample.conf [hunk 8]
```diff
@@ -0,0 +1,2 @@
+executable: gradle
```
Comment (blindpirate) [inline subprojects/docs/src/samples/ivy-publish/java-multi-project/publishing_ivy:publish-custom-artifact-snippet.sample.conf]: Well, this file name contains ":" and broke some windows tests... https://builds.gradle.org/viewLog.html?buildId=13947398&buildTypeId=Gradle_Check_Platform_Java8_Oracle_Windows_integTest

Comment (blindpirate) [reply subprojects/docs/src/samples/ivy-publish/java-multi-project/publishing_ivy:publish-custom-artifact-snippet.sample.conf]: https://github.com/gradle/gradle-private/issues/1353
